### ISSUE TITLE
NEON doesn't get enabled on ARM

### ISSUE THREAD

**Issue by tomoaki0705** [opencv/opencv#8945]
##### System information (version)

- OpenCV => master ( b21b6944447c95e364433c924dd560a16b91bffb )
- Operating System / Platform => Linux ARM32bit / Linux ARM64bit
- Compiler => GCC 4.8.4, 5.4.0 

##### Detailed description
 * CPU dispatch option is introduced, but it keeps blocking to activate the NEON on ARM platform
 * The main problem comes from obsolete options ```ENABLE_NEON``` [here](https://github.com/tomoaki0705/opencv/blob/b21b6944447c95e364433c924dd560a16b91bffb/CMakeLists.txt#L305-L306
)
```cmake
OCV_OPTION(ENABLE_NEON                "Enable NEON instructions"                                 "${NEON}" IF CMAKE_COMPILER_IS_GNUCXX AND (ARM OR AARCH64 OR IOS) )
OCV_OPTION(ENABLE_VFPV3               "Enable VFPv3-D32 instructions"                            OFF  IF CMAKE_COMPILER_IS_GNUCXX AND (ARM OR AARCH64 OR IOS) )
```
 * and [here](https://github.com/tomoaki0705/opencv/blob/b21b6944447c95e364433c924dd560a16b91bffb/cmake/OpenCVCompilerOptimizations.cmake#L79-L80)
```cmake
ocv_optimization_process_obsolete_option(ENABLE_VFPV3 VFPV3 OFF)
ocv_optimization_process_obsolete_option(ENABLE_NEON NEON OFF)
```
 * This macro ```ocv_optimization_process_obsolete_option``` will check the obsolete option and pass it to the "new" dispatch option, but I think [there is a bug in this macro ](https://github.com/tomoaki0705/opencv/blob/b21b6944447c95e364433c924dd560a16b91bffb/cmake/OpenCVCompilerOptimizations.cmake#L57 ) which ends up always switching flags to OFF 
```cmake
    if(${legacy_flag}) # !!!! this will not be evaluated as ON/OFF !!!!
      if(NOT ";${CPU_BASELINE_REQUIRE};" MATCHES ";${OPT};")
        set(CPU_BASELINE_REQUIRE "${CPU_BASELINE_REQUIRE};${OPT}" CACHE STRING "${HELP_CPU_BASELINE_REQUIRE}" FORCE)
      endif()
    else()
      if(NOT ";${CPU_BASELINE_DISABLE};" MATCHES ";${OPT};")
        set(CPU_BASELINE_DISABLE "${CPU_BASELINE_DISABLE};${OPT}" CACHE STRING "${HELP_CPU_BASELINE_DISABLE}" FORCE)
      endif()
    endif()
```
 * as a conclusion, the ```NEON``` flag and ```FP16``` flags always become false because those flag always go to ```CPU_BASELINE_DISABLE``` option.
 * I'll push a fix for this issue.

##### Steps to reproduce
 * on ARM 32bit (Jetson TK1)
```
cmake -DCPU_BASELINE=NEON ..
```
 * on ARM 64bit (Jetson TX1)
```
cmake ..
```

**Comment by alalek** [comment]
Thank for report! Could you try #8900 ?

**Comment by alalek** [comment]
> if(${legacy_flag}) # !!!! this will not be evaluated as ON/OFF !!!!

Could you please add information about CMake version?

**Comment by tomoaki0705** [comment]
I should have checked the PRs more frequently.

I tried on Jetson TK1,
```
--     CMake:                       2.8.12.2
```
and TX1
```
--     CMake:                       3.5.1
```

I'll get back to you with the result from #8900 once I recover the access to my build machine.
Best.

**Comment by tomoaki0705** [comment]
So, it seems I'm basically wrong.

I checked on Jetson TK1 and TX1 again, and it seems that ```${legacy_flag}``` was evaluated correctly as ON/OFF.
I mixed up with the behavior of obsolete option ```ENABLE_NEON=OFF```

Using #8900, I tried following combinations and all went fine

platform|cmake option|result
---|---|---
ARM64bit|None|```Baseline: NEON FP16```
ARM32bit|None|```Baseline: NEON```
ARM32bit|```-DCPU_BASELINE=FP16```|```Baseline: FP16```

The only curious point is about the relation between ```FP16``` flag and ```NEON``` flag.
It's bit weird that ```Baseline: FP16``` occurs. ```FP16``` never exists without ```NEON```.
I'll add a comment to #8900 about this point.
About this PR, I'll close.

platform name|compiler|cmake
---|---|---
Jetson TK1|GCC 4.8.4|cmake 2.8.12.2
Jetson TX1|GCC 5.4.0|cmake 3.5.1
ODROID-C2|GCC 5.4.0|cmake 3.5.1
ODROID-X2|GCC 4.8.4|cmake 2.8.12.2
Raspberry Pi 3|GCC 4.9.2|cmake 3.6.2
